# Thông tin
1. **MSSV:** 22521586
2. **Họ Tên:** Trần Võ Lâm Trường
3. **Bài tập:** Bài tập 3
4. **Ngày thực hiện:** 26/3/2024  



# Kết nối đến google drive

In [ ]:
from google.colab import drive
drive.mount('/gdrive')

Drive already mounted at /gdrive; to attempt to forcibly remount, call drive.mount("/gdrive", force_remount=True).


# Đặt các đường dẫn cho bài tập

In [ ]:
# Khai báo đường dẫn đến môn học CS231.O21.2
import os
dir_monhoc = '/gdrive/My Drive/UIT/Semester 4/[CS231] Computer Vision/Assignment/CS231.O21.2'

# Khai báo đường dẫn đến bài tập
dir_baitap_3 = os.path.join (dir_monhoc, 'Baitap', 'Baitap3')
dir_data = os.path.join (dir_baitap_3, 'dogscats_small')

In [ ]:
dir_train = os.path.join (dir_data, 'train')
dir_test = os.path.join (dir_data, 'test')

# Khai báo thư viện

In [ ]:
import cv2 as cv
from google.colab.patches import cv2_imshow
import numpy as np
import matplotlib.pyplot as plt
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression


# Tính histogram cho ảnh

In [ ]:
def computeHistogram(pathfilename):
    img = cv.imread(pathfilename,0)
    hist = cv.calcHist([img],[0],None,
                       [256],[0,256])
    size = img.shape[0]*img.shape[1]
    hist = hist / size
    return hist

# Đọc ảnh

In [ ]:
def ReadData(path_to_files):
  features = []
  labels = []
  list_files = os.listdir(path_to_files)
  list_files.sort()
  for file_name in list_files:
      label = file_name.split(".")[0]
      img_path = os.path.join(path_to_files, file_name)
      hist = computeHistogram(img_path)
      features.append(hist.flatten())
      labels.append(label)
  return np.array(features), np.array(labels)

In [ ]:
train_features, train_labels = ReadData(dir_train)
test_features, test_labels = ReadData(dir_test)

# kNN

In [ ]:
model = KNeighborsClassifier(n_neighbors=1)
model.fit(train_features, train_labels)
acc = model.score(test_features, test_labels)
print("Accuracy:", acc)

Accuracy: 0.5666666666666667


# Logistic Regression

In [ ]:
# Khởi tạo mô hình Logistic Regression
model_1 = LogisticRegression()#random_state = 22521586)

# Huấn luyện mô hình với dữ liệu huấn luyện
model_1.fit(train_features, train_labels)

# Đánh giá độ chính xác trên dữ liệu thử nghiệm
acc_1 = model_1.score(test_features, test_labels)
print("Accuracy:", acc_1)

Accuracy: 0.36666666666666664


# So sánh 2 phương pháp kNN và Logistic Regression

## **Kết quả:**

* kNN: 0.5666666666666667
* LogisticRegression: 0.36666666666666664

## **Phân tích:**
**Dựa vào kết quả trên, ta có thể thấy:**
* Phương pháp kNN có độ chính xác cao hơn LogisticRegression trong việc phân loại ảnh chó và mèo trong tập dữ liệu dogscats_small.
* Độ chính xác của kNN cao hơn 20% so với LogisticRegression.

**Lý do:**

kNN phù hợp hơn cho tập dữ liệu dogscats_small:
* Tập dữ liệu này có kích thước nhỏ, chỉ với 80 ảnh.
* kNN hoạt động hiệu quả với tập dữ liệu nhỏ.
* kNN không yêu cầu mô hình phức tạp để học tập.

LogisticRegression có thể gặp khó khăn với tập dữ liệu nhỏ:
* LogisticRegression là mô hình học máy tuyến tính.
* Mô hình tuyến tính có thể gặp khó khăn trong việc học các mối quan hệ phức tạp trong tập dữ liệu nhỏ.